### PR TITLE
Add guide on operators and observables

### LINKED ISSUE #3575: Docs: Quantum operator representations
I think it would be great for our users to have a page or extend some of our pages in our documentation describing the different kind of representations we have for operators and what features each have and how they interrelate. 

Currently the list would be: 
- Unitary operators: gate/operation + unitary protocol 
- Quantum channels: 
    - channel protocol - good for quantum trajectory simulations and noise models 
    - mixture protocol for unitary mixtures - which is good for performance as the probabilities are not state dependent  
- Measurement: 
    - MeasurementGate 
    - [future] Channel with measurement keys 
- Observables: 
    - PauliString - for pure paulis, that are closer to experimentally viable measurements
    - PauliSum - for convenience of summing up expectation values from PauliStrings, also to represent arbitrary operators in the Pauli basis
    - Observable measurement framework 

And then if we introduce butterfly (Dirac?) operators - they would be a sparse representation for arbitrary operators relying on the Dirac formalism, where you can capture things like projectors. 

We should also capture where there are nice interoperability between these representations - e.g. PauliSum and PauliString and Dirac operators can be added and multiplied together.

### PR DESCRIPTION
**Pull request by rmlarose**
Fixes #3575. (Perhaps #1550 as well.)

As mentioned in the notebook, the intended audience for this guide is someone already familiar with these concepts, just not how they are defined/used in Cirq.

Let me know @balopat how close this is to what you had in mind! Comments and feedback of course welcome.

### CODE REVIEW

**Comment by rmlarose** [comment]
Now ready for review.

I learned that `PauliSumCollector.collect` which uses asyncio doesn't always mix well with notebooks - throws `RuntimeError: This event loop is already running` or a more mysterious papermill error (`KeyError: 'language_info'`) - but `collect_async` seems to work fine.

File: docs/operators_and_observables.ipynb [hunk 1]
```diff
@@ -0,0 +1,953 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "cellView": "form",
+    "id": "KQa9t_gadIuR"
+   },
+   "outputs": [],
+   "source": [
+    "#@title Licensed under the Apache License, Version 2.0 (the \"License\");\n",
```
Comment (vtomole) [inline docs/operators_and_observables.ipynb:21]: Remove this `@title`

**Review by vtomole — CHANGES_REQUESTED** [review]

File: docs/operators_and_observables.ipynb (hunk 1 above)
Comment (balopat) [reply docs/operators_and_observables.ipynb:21]: We need that for the site!

Comment (vtomole) [reply docs/operators_and_observables.ipynb:21]: Seems like I've been super out of the loop when it comes to docs lately.

**Review by vtomole — APPROVED** [review]
This is fine to me.

**Review by balopat — CHANGES_REQUESTED** [review]
Thanks for this, it's a great start, good to see a comprehensive overview on this topic! I added some initial comments. 

I wonder if we want to talk about projectors (`cirq.KET_IMAG.projector()`) here too?

File: docs/operators_and_observables.ipynb [hunk 2]
```diff
@@ -0,0 +1,953 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "330de6a68bb9"
+   },
+   "outputs": [],
+   "source": [
+    "qbit = cirq.LineQubit(0)\n",
```
[the remark this answers is not in the document]
Comment (rmlarose) [reply docs/operators_and_observables.ipynb]: Updated here and throughout.

File: docs/operators_and_observables.ipynb [hunk 3]
```diff
@@ -0,0 +1,953 @@
...
+    "    <th>Object or Example</th>\n",
+    "    <th>Dedicated Guide(s)</th>\n",
+    "  </tr>\n",
+    "</thead>\n",
+    "<tbody>\n",
+    "  <tr>\n",
+    "    <td rowspan=\"3\">Operators</td>\n",
+    "    <td>Unitary operators</td>\n",
+    "    <td>`cirq.X`</td>\n",
+    "    <td>[Gates](gates.ipynb) and [Custom gates](custom_gates.ipynb)</td>\n",
```
[the remark this answers is not in the document]
Comment (rmlarose) [reply docs/operators_and_observables.ipynb]: Changed to markdown. (Only reason for html was rowspan which isn't supported by markdown, but this isn't needed with the new table as below.)

File: docs/operators_and_observables.ipynb [hunk 4]
```diff
@@ -0,0 +1,953 @@
...
+    "## Operators"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "7cabbfa3ba55"
+   },
+   "source": [
+    "Quantum operations (or just *operators*) include unitary gates, measurements, and noisy channels. Every operator in Cirq is an instance of a `cirq.Operation` which supports the Kraus operator representation\n",
```
[the remark this answers is not in the document]
Comment (rmlarose) [reply docs/operators_and_observables.ipynb]: Used this wording here and throughout.

File: docs/operators_and_observables.ipynb [hunk 5]
```diff
@@ -0,0 +1,953 @@
...
+    "    <td>`cirq.PauliString`</td>\n",
+    "    <td></td>\n",
+    "  </tr>\n",
+    "  <tr>\n",
+    "    <td>Hamiltonians</td>\n",
+    "    <td>`cirq.PauliSum`</td>\n",
+    "    <td></td>\n",
+    "  </tr>\n",
+    "</tbody>\n",
+    "</table>"
```
[the remark this answers is not in the document]
Comment (rmlarose) [reply docs/operators_and_observables.ipynb]: Thanks! I like this and adopted it almost verbatim with just a few cosmetic changes.

**Comment by rmlarose** [comment]
> I wonder if we want to talk about projectors (`cirq.KET_IMAG.projector()`) here too?

Seems like a good idea - probably in the section on measurements? I'm thinking of a simple example showing measurements in a different basis, after describing `cirq.measure_state_vector` and `cirq.measure_density_matrix` which measure in the `Z` basis. Something like

```python
prob_plus = np.trace(cirq.KET_PLUS.projector() @ rho).real
```

where `rho` is the same single-qubit density matrix from the previous example. Or maybe a two-qubit example is better. Is this what you had in mind?

**Comment by mpharrigan** [comment]
How did you fix KeyError: 'language_info' ?

**Comment by rmlarose** [comment]
> How did you fix KeyError: 'language_info' ?

Just to document for posterity: Looks like this was unrelated to asyncio as it popped up in #3669, see also https://github.com/quantumlib/Cirq/issues/3603#issuecomment-765830201. So here I'm not sure why the error went away.

**Comment by balopat** [comment]
> > I wonder if we want to talk about projectors (`cirq.KET_IMAG.projector()`) here too?
> 
> Seems like a good idea - probably in the section on measurements? I'm thinking of a simple example showing measurements in a different basis, after describing `cirq.measure_state_vector` and `cirq.measure_density_matrix` which measure in the `Z` basis. Something like
> 
> ```python
> prob_plus = np.trace(cirq.KET_PLUS.projector() @ rho).real
> ```
> 
> where `rho` is the same single-qubit density matrix from the previous example. Or maybe a two-qubit example is better. Is this what you had in mind?

Hmm...I was just thinking about a separate section for Projector operators, which are also coming in form of `KetBraSum` operators soon(ish) #3386. Simulating measurement this way would be an interesting idea too! 

I think this all is looking really good now. We can add the projectors as part of a separate issue, after we implemented the `KetBraSum` features. Then we can expand on pros/cons there - in the end `KetBraSum` should be a more generic and efficient, sparse representation for projectors.

**Review by balopat — APPROVED** [review]

**Comment by rmlarose** [comment]
> We can add the projectors as part of a separate issue, after we implemented the `KetBraSum` features. Then we can expand on pros/cons there - in the end `KetBraSum` should be a more generic and efficient, sparse representation for projectors.

Sounds good!

**Pull request merged**